gh-132962: _pyrepl: Prevent crash on Windows when stdout is redirected

shuimu5418 opened the pull request:
The interactive interpreter (-i) on Windows no longer crashes when standard output is redirected. The check for `CAN_USE_PYREPL` now also checks if stdout is a TTY on Windows, falling back to the basic REPL if it is not.

<!-- gh-issue-number: gh-132962 -->
* Issue: gh-132962
<!-- /gh-issue-number -->

---

Main branch:
```bash
PS F:\github\cpython001> .\python.bat -i .\build\testpys\print_helloworld.py > .\build\file.txt
Traceback (most recent call last):
  File "F:\github\cpython001\Lib\_pyrepl\readline.py", line 394, in multiline_input
    return reader.readline()
           ~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\reader.py", line 742, in readline
    self.prepare()
    ~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\historical_reader.py", line 306, in prepare
    super().prepare()
    ~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\reader.py", line 585, in prepare
    self.console.prepare()
    ~~~~~~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\windows_console.py", line 358, in prepare
    self.height, self.width = self.getheightwidth()
                              ~~~~~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\windows_console.py", line 409, in getheightwidth
    raise WinError(GetLastError())
OSError: [WinError 0] The operation completed successfully.
Traceback (most recent call last):
  File "F:\github\cpython001\Lib\_pyrepl\readline.py", line 394, in multiline_input
    return reader.readline()
           ~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\reader.py", line 742, in readline
    self.prepare()
    ~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\historical_reader.py", line 306, in prepare
    super().prepare()
    ~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\reader.py", line 585, in prepare
    self.console.prepare()
    ~~~~~~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\windows_console.py", line 358, in prepare
    self.height, self.width = self.getheightwidth()
                              ~~~~~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\windows_console.py", line 409, in getheightwidth
    raise WinError(GetLastError())
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect.
Traceback (most recent call last):
  File "F:\github\cpython001\Lib\_pyrepl\readline.py", line 394, in multiline_input
    return reader.readline()
           ~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\reader.py", line 742, in readline
    self.prepare()
    ~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\historical_reader.py", line 306, in prepare
    super().prepare()
    ~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\reader.py", line 585, in prepare
    self.console.prepare()
    ~~~~~~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\windows_console.py", line 358, in prepare
    self.height, self.width = self.getheightwidth()
                              ~~~~~~~~~~~~~~~~~~~^^
  File "F:\github\cpython001\Lib\_pyrepl\windows_console.py", line 409, in getheightwidth
    raise WinError(GetLastError())
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect.
Traceback (most recent call last):
...... Infinite Tracebacks......
```

With fix:
```bash
PS F:\github\cpython001> .\python.bat -i .\build\testpys\print_helloworld.py > .\build\file.txt
warning: can't use pyrepl: [Errno 25] tty required: 'stdout'
>>> 
```

shuimu5418 committed:
gh-132962: _pyrepl: Prevent crash on Windows when stdout is redirected

The interactive interpreter (-i) on Windows no longer crashes when
standard output is redirected. The check for `CAN_USE_PYREPL` now
also checks if stdout is a TTY on Windows, falling back to the
basic REPL if it is not.

ZeroIntensity reviewed (commented):
Please add a test case. You can refer to the [devguide](https://devguide.python.org/testing/run-write-tests/) or tag me if you're unsure what that entails.

shuimu5418 committed:
Add a test case to test_windows_console.py

shuimu5418 committed:
Replace sleep() with timeout in test case

shuimu5418 committed:
Fix formatting errors and make it pass CI tests

pablogsal committed:
gh-135371: Fix asyncio introspection output to include internal coroutine chains (#135436)

sobolevn committed:
gh-135504: Document `LIBZSTD_CFLAGS` and `LIBZSTD_LIBS` config options (#135505)

alebcay committed:
gh-135497: fix `MAXLOGNAME` detection in `configure.ac` (#135508)

serhiy-storchaka committed:
gh-135171: Update documentation for the generator expression (GH-135351)

* gh-135171: Update documentation for the generator expression

Document that the iterator for the leftmost "for" clause is created
immediately.

* Update Doc/reference/expressions.rst

Co-authored-by: Brian Skinn <brian.skinn@gmail.com>

---------

Co-authored-by: Brian Skinn <brian.skinn@gmail.com>

jeremycline committed:
gh-127319: Disable port reuse on HTTP, XMLRPC, and logging TCP servers (GH-135405)

Prior to issue #120485 these servers did not allow port reuse, which
makes sense as the behavior of port reuse is surprising if you're not
expecting it. It's unclear to me why these services were switched to
allow port reuse, but I believe the desired behavior (unless subclasses
opt in) is to not allow port reuse.

See also: https://bugzilla.redhat.com/show_bug.cgi?id=2323170

ZeroIntensity commented on Lib/test/test_pyrepl/test_windows_console.py:
```diff
@@ -577,5 +580,34 @@ def test_up_vt(self):
+                    "OSError" in stderr_output and
+                    len(stderr_output) > 1200
+                )
+                if has_crash_traceback:
```
As I said earlier, we should avoid exercising the bug, and instead exercise the correct behavior. There's no need for explicit detection of the infinite traceback. 

We should just call `process.wait()` without a timeout, and then assert that `script has run` is in the stdout.

shuimu5418 replied:
We used the -i parameter, so if proc.wait() is called, Python will wait forever. There's not much I can do about this. However, exercising the correct behavior is absolutely possible. Thanks.

lincolnj1 committed:
gh-65697: Improved error msg for configparser key validation (#135527)

* Improved error msg for configparser key validation and added note in 3.14 whatsnew

* Properly added change to configparser

* 📜🤖 Added by blurb_it.

---------

Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>

pablogsal committed:
gh-135371: Clean tags from pointers in all cases in remote debugging module (#135534)

Zheaoli committed:
gh-135361: update documentation for `remote_debugger_script` audit event (#135362)

picnixz committed:
gh-111178: fix UBSan failures for `RemoteUnwinderObject` (#135539)

dlenski committed:
gh-67022: Document bytes/str inconsistency in email.header.decode_header() and suggest email.headerregistry.HeaderRegistry as a sane alternative (#92900)

* gh-67022: Document bytes/str inconsistency in email.header.decode_header()

This function's possible return types have been surprising and error-prone
for the entirety of its Python 3.x history. It can return either:

1. `typing.List[typing.Tuple[bytes, typing.Optional[str]]]` of length >1
2. or `typing.List[typing.Tuple[str, None]]`, of length exactly 1

This means that any user of this function must be prepared to accept either
`bytes` or `str` for the first member of the 2-tuples it returns, which is a
very surprising behavior in Python 3.x, particularly given that the second
member of the tuple is supposed to represent the charset/encoding of the
first member.

This patch documents the behavior of this function, and adds test cases
to demonstrate it.

As discussed in bpo-22833, this cannot be changed in a backwards-compatible
way, and some users of this function depend precisely on the existing
behavior.

Add warnings about obsolescence of 'email.header.decode_header' and 'email.header.make_header' functions.

Recommend use of `email.headerregistry.HeaderRegistry` instead, as suggested
in https://github.com/python/cpython/pull/92900#discussion_r1112472177

shuimu5418 committed:
Fix test

shuimu5418 committed:
Fix CI

shuimu5418 committed:
Merge branch 'main' into fix-gh-132962-win-repl-redirect
